lib/source-map/quick-sort.js

```diff
@@ -0,0 +1,128 @@
+/* -*- Mode: js; js-indent-level: 2; -*- */
+/*
+ * Copyright 2011 Mozilla Foundation and contributors
+ * Licensed under the New BSD license. See LICENSE or:
+ * http://opensource.org/licenses/BSD-3-Clause
+ */
+if (typeof define !== 'function') {
+    var define = require('amdefine')(module, require);
+}
+define(function (require, exports, module) {
+
+  // It turns out that some (most?) JavaScript engines don't self-host
+  // `Array.prototype.sort`. This makes sense because C++ will likely remain
+  // faster than JS when doing raw CPU-intensive sorting. However, when using a
+  // custom comparator function, calling back and forth between the VM's C++ and
+  // JIT'd JS is rather slow *and* loses JIT type information, resulting in
+  // worse generated code for the comparator function than would be optimal. In
+  // fact, when sorting with a comparator, these costs outweigh the benefits of
+  // sorting in C++. By using our own JS-implemented Quick Sort (below), we get
+  // a ~3500ms mean speed-up in `bench/bench.html`.
+
+  /**
+   * Swap the elements indexed by `x` and `y` in the array `ary`.
+   *
+   * @param {Array} ary
+   *        The array.
+   * @param {Number} x
+   *        The index of the first item.
+   * @param {Number} y
+   *        The index of the second item.
+   */
+  function swap(ary, x, y) {
+    var temp = ary[x];
+    ary[x] = ary[y];
+    ary[y] = temp;
+  }
+
+  /**
+   * Returns a random integer within the range `low .. high` inclusive.
+   *
+   * @param {Number} low
+   *        The lower bound on the range.
+   * @param {Number} high
+   *        The upper bound on the range.
+   */
+  function randomIntInRange(low, high) {
+    return Math.round(low + (Math.random() * (high - low)));
+  }
+
+  /**
+   * The partition function chooses a pivot between `p` and `r` and moves all
+   * elements that are less than or equal to the pivot to the before it, and all
+   * the elements that are greater than it after it. The effect is that once
+   * partition is run, the pivot is in the exact place it will be when the array
+   * is put in sorted order, and it will not need to be moved again. This runs
+   * in O(n) time.
+   *
+   * @param {Array} ary
+   *        The array to partition.
+   * @param {function} comparator
+   *        The function with which to compare items.
+   * @param {Number} p
+   *        Start index of the array.
+   * @param {Number} r
+   *        End index of the array.
+   */
+  function partition(ary, comparator, p, r) {
```

Comment:
Why don't you inline this function? It seems to be only used from one place. We could save the overhead of a single function call for each recursive step.